heartbeat/monitors/active/http/task.go

```diff
@@ -208,38 +207,46 @@ func execPing(
 	event *beat.Event,
 	client *http.Client,
 	req *http.Request,
-	body []byte,
+	reqBody []byte,
 	timeout time.Duration,
 	validator func(*http.Response) error,
+	responseConfig responseConfig,
 ) (start, end time.Time, errReason reason.Reason) {
 	ctx, cancel := context.WithTimeout(context.Background(), timeout)
 	defer cancel()
 
-	req = attachRequestBody(&ctx, req, body)
-	start, end, resp, errReason := execRequest(client, req, validator)
+	req = attachRequestBody(&ctx, req, reqBody)
 
-	if errReason == nil || errReason.Type() != "io" {
-		eventext.MergeEventFields(event, common.MapStr{"http": common.MapStr{
-			"rtt": common.MapStr{
-				"total": look.RTT(end.Sub(start)),
-			},
-		}})
-	}
+	// Send the HTTP request. We don't immediately return on error since
+	// we may want to add additional fields to contextualize the error.
+	start, resp, errReason := execRequest(client, req, validator)
 
-	if resp != nil {
-		eventext.MergeEventFields(event, common.MapStr{"http": common.MapStr{
-			"response": common.MapStr{"status_code": resp.StatusCode},
-		}})
+	// If we have no response object there probably was an IO error, we can skip the rest of the logic
+	// since that logic is for adding metadata relating to completed HTTP transactions that have errored
+	// in other ways
+	if resp == nil {
+		return start, time.Now(), errReason
 	}
 
-	if errReason != nil {
-		if resp != nil {
-			return start, end, errReason
-		}
-		return start, end, errReason
+	// Add response.status_code
+	eventext.MergeEventFields(event, common.MapStr{"http": common.MapStr{"response": common.MapStr{"status_code": resp.StatusCode}}})
```

Comment:
I wonder if you want to do that only after the handleRespBody to not have that also in the timing.

Reply:
The reason this is first is because `handleRespBody` can return an error causing an early exit. It would be a little awkward to inject this before the error check/return.

This is pretty fast code to execute, and given that we aren't writing real time code, the effect should be negligible.